app/models/cherrypick_task.rb

```diff
@@ -33,23 +33,23 @@ def control_positions(batch_id, num_plate, total_wells, num_control_wells, wells
 
     raise StandardError, 'More controls than free wells' if num_control_wells > total_available_positions
 
-    quotient = batch_id
-    size_region = (total_available_positions / num_control_wells)
-    regions = (wells_to_leave_free..(total_wells - 1)).each_slice(size_region).to_a
+    available_positions = (wells_to_leave_free..(total_wells - 1)).to_a
+    # If num plate is equal to the available positions, the cycle is going to be repeated.
+    # To avoid it, every num_plate=available_positions we start a new cycle with a new seed.
+    seed = batch_id * ((num_plate / available_positions.length) + 1)
+    initial_positions = random_elements_from_list(available_positions, num_control_wells, seed)
+    control_positions_for_plate(num_plate, initial_positions, available_positions)
+  end
 
-    # Number of regions should equal number of controls, so sometimes last region is bigger than average
-    if regions.length > num_control_wells
-      last_region = regions.slice!(regions.length - 1, 1)
-      regions[regions.length - 1].concat(*last_region)
-    end
+  def random_elements_from_list(list, num_elems, seed)
+    list.sample(num_elems, random: Random.new(seed))
+  end
+
+  def control_positions_for_plate(num_plate, initial_positions, available_positions)
+    return initial_positions if num_plate.zero?
 
-    position = 0
-    regions.each_with_index.map do |region, _num_region|
-      quotient, remain = quotient.divmod(region.length)
-      # Feeding the new position with the old position seems to add more
-      # variability between the controls positions
-      position = (remain + num_plate + position) % region.length
-      region[position]
+    initial_positions.map do |pos|
+      available_positions[(available_positions.index(pos) + num_plate) % available_positions.length]
```

Comment:
`available_positions[(available_positions.index(pos) + num_plate) % available_positions.length]`
Can now just become:
`available_positions[(available_positions.index(pos)]` as available positions is already randomized based on plate number. (This is still the case when you make the change I suggested above)

Reply:
Actually, I'm not even sure you need that much, I think you can just use initial positions?

Reply:
Yes, that was to apply the logic of increasing with num plate, so depends on how each plate is generated, but if it is random, then it can be initial_positions (although then that variable name will need rewording)